Treating onabort as an error

schmidt-sebastian opened the pull request:
This PR removes our special handling of `onabort` for IndexedDb transactions and treats aborts as an error (as `onabort` is called for quota exceeded). We currently don't abort transactions in our SDK and hence removing this handling seems like the proper thing to do.

I tested this in a test app (not using the Firestore SDK) and saw this error get set during both the initial schema migration as well as during normal operation. The error is a DOMException with `{ code: 22, name: "QuotaExceededError" }`. To get there, I had to write 32 GB of random data to IndexedDb :)

The 'still propagates error if you throw after aborting an exception' test no longer makes sense since the abort already rejects the transaction.

Fixes: https://github.com/firebase/firebase-js-sdk/issues/720

mikelehen commented on packages/firestore/test/unit/local/simple_db.test.ts:
```diff
@@ -128,11 +133,13 @@ describe('SimpleDb', () => {
+          txn.abort();
+        });
+      })
+    ).to.eventually.be.rejectedWith('The IndexedDb transaction was aborted.');
```
FWIW, I don't think aborting a transaction should generally be an error case, so I'd maybe add a todo...

// TODO(mrschmidt): If in the future we end up adding code that intentionally aborts transactions we may need to turn this into a non-error case.

schmidt-sebastian replied:
Done.

mikelehen replied:
Per my other comment, I think we can make this work again without generating an error?

schmidt-sebastian replied:
Yeah, with the conditional check for `transaction.error`, this once again becomes possible.

mikelehen commented on yarn.lock:
```diff
@@ -233,10 +233,10 @@
 "@types/body-parser@*":
-  version "1.16.8"
-  resolved "https://registry.npmjs.org/@types/body-parser/-/body-parser-1.16.8.tgz#687ec34140624a3bec2b1a8ea9268478ae8f3be3"
+  version "1.17.0"
```
uh, wow...  there was a lot of churn in this file.  All of that is from adding chai-as-promised, right?

schmidt-sebastian replied:
This took all morning to resolve, but I am now happy to report that the diff is now as expected.

schmidt-sebastian commented:
> By the way, from your comment I can't 100% tell if you verified that in the case of a quota exceeded error enablePersistence() will now return a rejected promise but leave the client functional (just with persistence disabled)... Can you confirm?

We need to special-case this error in `canRecover()`, which is something that I hadn't done. I updated the PR.

I also fixed an issue with our error bubbling. Before, we would call `abort()`, which then rejected the Promise before we had a chance to reject it with the error that we threw.

mikelehen commented on packages/firestore/src/local/simple_db.ts:
```diff
@@ -263,19 +260,30 @@ export class SimpleDbTransaction {
+    }
+
     if (!this.aborted) {
       debug(LOG_TAG, 'Aborting transaction.');
```
perhaps log the error?

schmidt-sebastian replied:
Done

mikelehen commented on packages/firestore/src/local/simple_db.ts:
```diff
@@ -147,15 +148,14 @@ export class SimpleDb {
         // error.
-        transaction.abort();
-        return PersistencePromise.reject(error);
+        transaction.abort(error);
```
Can you rewrap the comment while you're here?

schmidt-sebastian replied:
Done

mikelehen commented on packages/firestore/src/local/simple_db.ts:
```diff
@@ -246,11 +246,8 @@ export class SimpleDbTransaction {
-   * and this promise will resolve as successful.
    */
-  readonly completionPromise: Promise<void>;
+  readonly completionPromise: Deferred<void>;
```
Err...   exposing the raw Deferred<> as part of the public API of SimpleDbTransaction seems like leaking implementation details. External code shouldn't be able to call `transaction.completionPromise.resolve()` and external code shouldn't have to do `transaction.completionPromise.promise` to get the promise.  I would:

1. Rename completionPromise to completionDeferred and make it private.
2. Add a public completionPromise getter that returns completionDeferred.promise.

schmidt-sebastian replied:
Done and done.

mikelehen commented on packages/firestore/src/local/simple_db.ts:
```diff
@@ -263,19 +260,30 @@ export class SimpleDbTransaction {
+      const error =
+        transaction.error ||
+        new Error('The IndexedDb transaction was aborted.');
+      this.completionPromise.reject(error);
```
So actually...  could we change this to:

```
if (transaction.error) {
  this.completionPromise.reject(error);
}
```
?

I think that would mean:
1. You can call transaction.abort() [no error] to gracefully abort the transaction but not trigger any error handling.
2. It would solve a wart where as-is, I think if you call transaction.abort(error), then we'll end up calling completionPromise.reject() in abort() and then again here which seems sloppy / confusing.

schmidt-sebastian replied:
Done

mikelehen commented on packages/firestore/test/unit/local/simple_db.test.ts:
```diff
@@ -142,18 +149,13 @@ describe('SimpleDb', () => {
-    }
-    expect(gotError).to.equal(true);
+      })
+    ).to.eventually.be.rejectedWith('Generated error');
```
FWIW, looking at the chai-as-promised examples, they don't tend to use expect() so this would become:

```
await runTransaction(...).should.eventually.be.rejectedWith('Generated error');
```

I don't care strongly though, so feel free to leave as-is.

schmidt-sebastian replied:
I changed this to use `expect` since I found it somewhat questionable to just extend the Promise implementation. Using `expect` hides this a little bit.

schmidt-sebastian commented on packages/firestore/src/core/firestore_client.ts:
```diff
@@ -226,11 +229,20 @@ export class FirestoreClient {
-      error.code === Code.FAILED_PRECONDITION ||
-      error.code === Code.UNIMPLEMENTED
-    );
+  private canFallback(error: FirestoreError | DOMException): boolean {
```
Done

schmidt-sebastian commented on packages/firestore/src/core/firestore_client.ts:
```diff
@@ -56,6 +56,9 @@ import { ViewSnapshot } from './view_snapshot';
 const LOG_TAG = 'FirestoreClient';
 
+/** The DOMException code for quota exceeded. */
+const QUOTA_EXCEEDED = 22;
```
Done.

mikelehen commented on packages/firestore/src/local/simple_db.ts:
```diff
@@ -263,21 +259,36 @@ export class SimpleDbTransaction {
-      debug(LOG_TAG, 'Aborting transaction.');
+      debug(
+        LOG_TAG,
+        'Aborting transaction: %s',
```
Uhh...  Does `%s` work in JavaScript?

schmidt-sebastian replied:
It does :)
